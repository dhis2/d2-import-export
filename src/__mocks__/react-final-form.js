export const Form = ({ children }) => <Fragment>{children}</Fragment>

export const useField = jest.fn(() => ({
    input: {
        name: 'Name',
        value: '',
        onChange: jest.fn(),
    },
    meta: {
        touched: false,
        error: '',
    },
}))
