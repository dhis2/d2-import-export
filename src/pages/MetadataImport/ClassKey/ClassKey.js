import React, { useState } from 'react'
import PropTypes from 'prop-types'
import i18n from '@dhis2/d2-i18n'
import { useDataQuery } from '@dhis2/app-runtime'

import { Select } from '../../../components/Select'

const classKeyQuery = {
    keys: {
        resource: 'metadata/csvImportClasses',
    },
}

const ClassKey = ({ prevValue, value, setValue, dataTest }) => {
    const [error, setError] = useState(undefined)
    const [classKeyOptions, setClassKeyOptions] = useState([])

    const { loading } = useDataQuery(classKeyQuery, {
        onComplete: classData => {
            setClassKeyOptions(
                classData.keys.map(k => ({ value: k, label: k }))
            )
            if (prevValue && !value) {
                setValue(prevValue)
            } else if (!value) {
                setValue({
                    value: classData.keys[0],
                    label: classData.keys[0],
                })
            }
        },
        onError: error => {
            setError(error)
            console.error('ClassKey error: ', error)
        },
    })

    const validationText =
        error &&
        `${i18n.t('Something went wrong when loading the class keys')} : ${
            error.message
        }`

    return (
        <Select
            name="classKey"
            label={i18n.t('Class key')}
            options={classKeyOptions}
            selected={loading ? undefined : value}
            setValue={setValue}
            loading={loading}
            dataTest={dataTest}
            validationText={validationText}
            error={!!error}
            dense
        />
    )
}

ClassKey.propTypes = {
    dataTest: PropTypes.string.isRequired,
    setValue: PropTypes.func.isRequired,
    prevValue: PropTypes.object,
    value: PropTypes.object,
}

export { ClassKey }
