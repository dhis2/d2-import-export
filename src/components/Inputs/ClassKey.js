import { useForm } from 'react-final-form'
import React, { Fragment } from 'react'
import i18n from '@dhis2/d2-i18n'
import propTypes from 'prop-types'

import { Field } from '../Field/Field'
import { RadioGroup } from '../FinalFormComponents/RadioGroup'
import { Select } from '../FinalFormComponents/Select'

export const CLASS_KEY_KEY = 'classKey'

export const ClassKey = ({ show, options, defaultValue }) => {
    const { change, getState } = useForm()
    const currentValue = getState().values[CLASS_KEY_KEY]

    // not directly imported so the `useEffect` hook
    // can be stubbed so the callback is not called async'ly
    React.useEffect(() => {
        if (!show && currentValue) {
            change(CLASS_KEY_KEY, null)
        }
    }, [change, show, currentValue])

    return (
        <Fragment>
            {show && (
                <Field>
                    <Select
                        dataTest="input-class-key"
                        name={CLASS_KEY_KEY}
                        label={i18n.t('Class key')}
                        options={options}
                        defaultValue={defaultValue}
                    />
                </Field>
            )}
        </Fragment>
    )
}

ClassKey.propTypes = {
    options: RadioGroup.propTypes.options,
    show: propTypes.bool.isRequired,
    defaultValue: propTypes.string.isRequired,
}
