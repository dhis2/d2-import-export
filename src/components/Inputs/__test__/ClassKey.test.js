import { useForm } from 'react-final-form'
import React from 'react'

import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import { CLASS_KEY_KEY, ClassKey } from '../ClassKey'

jest.mock('react-final-form')

describe('Input component - ClassKey', () => {
    React.useEffect = jest.fn(callback => callback())

    const defaultValue = 'foo'
    const options = [
        { value: 'foo', label: 'Foo' },
        { value: 'bar', label: 'Bar' },
    ]

    beforeEach(() => {
        useForm.mockClear()
        React.useEffect.mockClear()
    })

    it('should render correctly', () => {
        useForm.mockImplementationOnce(() => ({
            change: jest.fn(),
            getState: () => ({ values: { [CLASS_KEY_KEY]: 'foo' } }),
        }))

        const file = shallow(
            <ClassKey
                show={true}
                options={options}
                defaultValue={defaultValue}
            />
        )
        expect(toJson(file)).toMatchSnapshot()
    })

    it('should remove the value from the form state when show is false and there is a value', () => {
        const change = jest.fn()
        useForm.mockImplementationOnce(() => ({
            change,
            getState: () => ({ values: { [CLASS_KEY_KEY]: 'foo' } }),
        }))

        const file = shallow(
            <ClassKey
                show={false}
                options={options}
                defaultValue={defaultValue}
            />
        )
        expect(change).toHaveBeenCalledTimes(1)
        expect(change).toHaveBeenCalledWith(CLASS_KEY_KEY, null)
    })
})
