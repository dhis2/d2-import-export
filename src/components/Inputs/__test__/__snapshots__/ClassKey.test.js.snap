// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component - ClassKey should render correctly 1`] = `
<Fragment>
  <Field>
    <Select
      dataTest="input-class-key"
      defaultValue="foo"
      label="Class key"
      name="classKey"
      options={
        Array [
          Object {
            "label": "Foo",
            "value": "foo",
          },
          Object {
            "label": "Bar",
            "value": "bar",
          },
        ]
      }
    />
  </Field>
</Fragment>
`;
